### PR TITLE
[skip ci] Implement structure and inplace coordinate functions

### PR DESCRIPTION
**Pull request by abelsiqueira**
Rebasing into `coord-and-structure` to handle #160. After all PRs are merged into `coord-and-structure`, I'll open a PR which runs CI. **This doesn't run CI**.

Part 01 of 07.

### CODE REVIEW

File: src/NLPModels.jl [hunk 1]
```diff
@@ -172,6 +173,19 @@ function objgrad!(nlp, x, g)
   return f, g
 end
 
+"""`(rows,cols) = jac_structure(nlp)`
```
Comment (dpo) [inline src/NLPModels.jl]: I think the calling sequence should be on the second line, indented 4 spaces and with no backticks.

File: src/NLPModels.jl [hunk 2]
```diff
@@ -172,6 +173,19 @@ function objgrad!(nlp, x, g)
...
+
+Returns the structure of the constraint's Jacobian in sparse coordinate format.
+"""
+jac_structure(:: AbstractNLPModel) = throw(NotImplementedError("jac_structure"))
+
+"""`(rows,cols,vals) = jac_coord!(nlp, x, rows, cols, vals)`
+
+Evaluate \$\\nabla c(x)\$, the constraint's Jacobian at `x` in sparse coordinate format,
```
Comment (dpo) [inline src/NLPModels.jl]: How does the LaTeX show in the console? Is it readable?

File: src/NLSModels.jl [hunk 3]
```diff
@@ -115,6 +117,25 @@ function jac_residual(nls :: AbstractNLSModel, x :: AbstractVector)
...
+  throw(NotImplementedError("jac_structure_residual"))
+end
+
+"""
+    (rows,cols,vals) = jac_coord_residual!(nls, x, rows, cols, vals)
+
+Computes the Jacobian of the residual at `x` in sparse coordinate format, rewriting
+`vals`. `rows` and `cols` are not rewritten.
+"""
+function jac_coord_residual!(nls :: AbstractNLSModel, x :: AbstractVector, rows :: AbstractVector{Int}, cols :: AbstractVector{Int}, vals :: AbstractVector)
```
Comment (dpo) [inline src/NLSModels.jl]: Does this work? I thought `Vector{Int} <: AbstractVector{Int}` was `false`?!

File: src/NLPModels.jl (hunk 2 above)
Comment (abelsiqueira) [reply src/NLPModels.jl]: Here, but `\nabla c(x)` is colored differently, with a different color that `x`, `vals`, etc.
```
  (rows,cols,vals) = jac_coord!(nlp, x, rows, cols, vals)

  Evaluate \nabla c(x), the constraint's Jacobian at x in sparse coordinate format, rewriting vals.
  rows and cols are not rewritten.
```

File: src/NLSModels.jl (hunk 3 above)
Comment (abelsiqueira) [reply src/NLSModels.jl]: `Vector{Int} <: AbstractVector{Int}` is true, but `AbstractVector{Int32} <: AbstractVector{Int}` is false.

File: src/NLPModels.jl (hunk 1 above)
Comment (abelsiqueira) [reply src/NLPModels.jl]: Most docstrings in `NLPModels.jl` are this way. I'll open an issue to change them afterwards.

Comment (dpo) [reply src/NLPModels.jl]: Ok

**Pull request merged**